src/source/native.rs

```diff
@@ -70,42 +69,59 @@ impl TCPStreamHandler for NativeStreamHandler {
         poller: &mio::Poll,
         stream: mio::net::TcpStream,
     ) -> () {
-        let mut reader = io::BufReader::new(stream);
-        loop {
+        let mut streaming = true;
+        let mut reader = BufferedPayload::new(stream.try_clone().unwrap());
+        while streaming {
             let mut events = mio::Events::with_capacity(1024);
             match poller.poll(&mut events, None) {
                 Err(e) => panic!(format!("Failed during poll {:?}", e)),
                 Ok(_num_events) => for event in events {
                     match event.token() {
                         constants::SYSTEM => return,
                         _stream_token => {
-                            let rchans = chans.clone();
-                            self.handle_stream_payload(rchans, tags, &mut reader);
+                            while streaming {
+                                match reader.read() {
+                                    Ok(mut raw) => {
+                                        let rchans = chans.clone();
+                                        self.handle_stream_payload(rchans, tags, &mut raw);
```

Comment:
This has error conditions that don't seem to be returned / modeled. The Avro source is a little more sophisticated in this regard. Should these native source flows also handle errors instead of silently dropping them?

Reply:
I will make a ticket to beef up native's error logging + internal telemetry.